### PR TITLE
DB Backup Script Fix

### LINKED ISSUE #3079: DB Backup Script Fix
**Description:**
It was recently discovered in #3004 that the `db_backup.py` script has a subtle error. The script correctly creates, and stores a backup. However, the script is using `VCAP_SERVICES` to get the DB connection information. In all environments except for `production` there are multiple database names in the RDS instance. The logic that selects the DB connection information in the script is selecting the first DB it comes across which should not be the case. The script should choose the correct DB for all environments every time. As a bandaid solution, #3004 implemented the backup logic using Django settings which is not what we want. The script should not rely on Django.

**Acceptance Criteria:** 

- [ ] Backup script creates backups of the correct DB in the RDS instance across all environments
- [ ] Script does not require Django to work successfully
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Remove Django settings usage throughout backup script
- [x] Update logic to select correct DB information in `get_system_values()`
- [x] Test the backup script in a deployed environment
- [x] Verify in the production environment that we have valid backups despite this error
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

N/A

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

N/A

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

N/A

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Updated script to not rely on Django settings
- Updated logic and settings to strictly use the DB URI
- Added some extra checking for correct cmdline args for restoring a db
Pull request closes #3079 

## How to Test
1. In a deployed env change the cron schedule on the backup task so that it will execute
2. Watch the log entries to see where the backup was stored
3. Grab the s3 url
4. Follow cloud.gov's guidance to get access to s3 from local machine with AWS CLI
5. Copy the backup to your machine and verify it's size is larger than 1Kb (e.g. the Raft env DB is ~43MB as of 8/7/2024)

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Backup script creates backups of the correct DB in the RDS instance across all environments
- [x] Script does not require Django to work successfully
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: scripts/deploy-backend.sh [hunk 1]
```diff
@@ -105,12 +105,16 @@ update_backend()
...
-    
+
     if [ "$CF_SPACE" = "tanf-prod" ]; then
       cf set-env "$CGAPPNAME_BACKEND" AV_SCAN_URL "http://tanf-prod-clamav-rest.apps.internal:9000/scan"
     else
       # Add environment varilables for clamav
       cf set-env "$CGAPPNAME_BACKEND" AV_SCAN_URL "http://tdp-clamav-nginx-$env.apps.internal:9000/scan"
+
+      # Add variable for dev/staging apps to know their DB name. Prod uses default AWS name.
+      cf unset-env "$CGAPPNAME_BACKEND" "APP_DB_NAME"
```
Comment (raftmsohani) [inline scripts/deploy-backend.sh:116]: Is this being set by settings?

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by jtimpe — APPROVED** [review]

File: scripts/deploy-backend.sh (hunk 1 above)
Comment (ADPennington) [reply scripts/deploy-backend.sh:116]: @elipe17 ⬆️

**Review by ADPennington — APPROVED** [review]
@elipe17 lgtm!  🚀 Testing steps ⬇️ 

1. In a deployed env change the cron schedule on the backup task so that it will execute

```shell=
Aug. 14, 2024, 12:51 p.m.	alexandra.pennington@acf.hhs.gov	django_celery_beat | crontab	52,55 12 * * * (m/h/dM/MY/d) UTC	Change	[{"changed": {"fields": ["Minute(s)", "Hour(s)"]}}]
```
2. Watch the log entries to see where the backup was stored

```shell=
2024-08-14T08:52:45.69-0400 [APP/PROC/WEB/0] ERR 2024-08-14 12:52:45,698 INFO db_backup.py::upload_file:L186 :  Successfully uploaded /tmp/tdp_db_qasp_backup.pg to s3://<redacted>/backup/tmp/tdp_db_qasp_backup.pg.

2024-08-14T08:55:46.23-0400 [APP/PROC/WEB/0] ERR 2024-08-14 12:55:46,235 INFO db_backup.py::upload_file:L186 :  Successfully uploaded /tmp/tdp_db_qasp_backup.pg to s3://<redacted>/backup/tmp/tdp_db_qasp_backup.pg.
```

3. Grab the s3 url

```shell=
s3://<redacted>/backup/tmp/backup.pg.
```

4. Follow cloud.gov's guidance to get access to s3 from local machine with AWS CLI

```shell=

apennington@HHS /
$ aws s3 ls  s3://<redacted>/backup/tmp/backup.pg  --recursive --profile TDP_qasp_staticfilesS3_20240228
2024-08-14 00:00:50  281573132 backup/tmp/backup.pg
```

5. Copy the backup to your machine and verify it's size is larger than 1Kb (e.g. the Raft env DB is ~43MB as of 8/7/2024)

```shell=
apennington@HHSLBDSWL73 MINGW64 /
$ aws s3 cp s3://<redacted>/backup/tmp/backup.pg ./qasp_db_backup.pg --profile TDP_qasp_staticfilesS3_20240228
download: s3://<redacted>/backup/tmp/backup.pg to .\qasp_db_backup.pg
```
Screenshot 2024-08-14 092202

File: tdrs-backend/tdpservice/scheduling/management/db_backup.py [hunk 2]
```diff
@@ -268,28 +248,47 @@ def get_database_credentials(database_uri):
...
     return [username, password, host, port, database_name]
 
-
-def main(argv, sys_values, system_user):
-    """Handle commandline args."""
-    arg_file = "/tmp/backup.pg"
-    arg_database = sys_values['DATABASE_URI']
+def get_opts(argv, db_name):
+    """Parse command line options."""
+    arg_file = f"/tmp/{db_name}_backup.pg"
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/scheduling/management/db_backup.py:253]: this has replaced our previous backup in prod (backup.pg) with the db_name hash. is this what was intended?

**Pull request merged**